functions/src/stories.ts

```diff
@@ -11,83 +11,161 @@ const index = algoliasearch(
+        language: translation.language,
+        title,
+        description,
+        tags,
+      };
       if (change.before.exists) {
-        await index.deleteBy({
-          filters: "translationID:" + translationID,
-        });
-      }
-      if (change.after.exists) {
-        const {title, description, tags} = change.after.data()!;
-        const story = await db
-            .doc(`stories/${storyID}`)
-            .get()
-            .then((doc) => doc.data()!);
-        const translation = await db
-            .doc(`stories/${storyID}/translations/${translationID}`)
-            .get()
-            .then((doc) => doc.data()!);
+        await index.partialUpdateObject(
+            {objectID: translationID, ...entry},
+            {createIfNotExists: true}
+        );
+      } else {
         await index.saveObject(
-        {
-          storyID,
-          storyAuthorID: story.metaData.authorId,
-          translationID,
-          translationAuthorID: translation.metaData.authorId,
-          language: translation.language,
-          title,
-          description,
-          tags,
-          likes: translation.metaData.likes,
-        } as storyRecord,
-        {autoGenerateObjectIDIfNotExist: true}
+            entry,
+            {autoGenerateObjectIDIfNotExist: true}
         );
       }
     });
 
-// TODO
+export const trackStoryUpdateTime = functions
+    .region("europe-central2")
+    .firestore.document(
+        "stories/{storyID}/translations/{translationID}/assets/{assetID}"
+    )
+    .onWrite(async (change, context) => {
+      if (!change.after.exists) return;
+      await storyDoc(context.params.storyID, context.params.translationID)
+          .update({
+            "metaData.lastUpdateAt": firestore.FieldValue.serverTimestamp(),
+          });
+    });
+
+export const updateStoryMeta = functions
+    .region("europe-central2")
+    .firestore.document(
+        "stories/{storyID}/translations/{translationID}"
+    )
+    .onWrite(async (change, context) => {
+      if (change.after.exists) {
+        const meta = change.after.data()?.metaData;
+        await index.partialUpdateObject({
+          objectID: context.params.translationID,
+          likes: meta.likes ?? 0,
+          views: meta.views ?? 0,
+          lastUpdateAt: meta.lastUpdateAt,
+          imageUrl: meta.imageUrl,
+        });
+      }
+    });
+
+export const indexTrendingStories = functions
+    .region("europe-central2")
+    .pubsub.schedule("every 72 hours")
+    .onRun(async () => {
+      await index.browseObjects<StoryRecord>({
+        query: "",
+        attributesToRetrieve: ["storyID"],
+        batch: async (batch) => {
+          for (const hit of batch) {
+            const {storyID, objectID} = hit;
+            const doc = storyDoc(storyID, objectID);
+            const meta = await doc.get().then((r) => r.data()!.metaData);
+            const trending = (meta.views ?? 0) - (meta.lastIndexedViews ?? 0);
```

Comment:
So, a story is trending if it got more views than it had 72 hours ago? How can that number even decrease? And why is the "trendiness" of a story depending only on itself, and not also comparing it to other stories?

Reply:
All the stories in that replica index are sorted by that `trending` field, so they are in order. As for the logic, I have the following:
1. Say, today we have two stories with 15  and 5 views correspondingly. They've just been added and indexed, right so their `lastIndexedViews ` are equal to 15 and 5 too.
2. After three days, the first has 19 views, while the second has 14. The delta (`trending` value) of the first is 19-15=4, and of the second is 14-5=9. Hense the second one goes up. And their `lastIndexedViews ` become 19 and 14 until the next round.